Clean up the networking stack (#295)
filecoin-project/specs: networking.md

# Filecoin Network

The filecoin network is built using libp2p building blocks, as transports and protocols, as well as some additional filecoin specific protocols as outlined in [Network Protocols](./network-protocols.md).

## Required Protocols

Every full node must support the following libp2p protocols:

 - [gossipsub](https://github.com/libp2p/specs/tree/master/pubsub/gossipsub) ([for data announcements](./data-propagation.md))

>  - [bitswap] ([for data exchange](./data-propagation.md))

Reviewer: Missing link to bitswap

Author: `[for data exchange](./data-propagation.md)` links generally don't seem to since the hugo setup anymore. @dignifiedquire any idea why or how to do it instead (without breaking regular markdown links?)

 - Filecoin specific Protocols:
   - Hello Handshake
   - StorageDeal

>    - BlockSync

Reviewer: What about the other protocols defined?

Reviewer 2: I'd link these into the relevant network-protocols.md portions

Author: as said, we have a linking issue, waiting for that to be resolved.

This document doesn't make any assumptions on which transports and multiplexing features a particular node offers, however it must at least be able to respond and communicate through mplex channels via a secio transport.

{{% notice info %}}
Implementers are encouraged to offer and experiment with more protocols and transports, as long as the above mentioned requirements of ee2e, authentication and muxing are still met. In particular the authors encourage to look into TLS1.3 and QUIC as libp2p protocol/transport layers.
{{% /notice %}}

### Establishing connections

When a filecoin node connects to the network, it may connect to multiple nodes at once. When a new node connects, it must be greeted with the `HelloMessage` through the hello-handshake protocol first. Only after that message has been sent can other streams be opened. If a node receives a `HelloMessage` with a GenesisBlock it doesn't support, it must immediately close the entire connection to that peer.

> If the node learns through that handshake about newer blocks it should request these immediately to sync up their chain.

Reviewer 3: I think the situation here may actually be a bit more complex: generally, some "newer blocks" may be invalid (not in themselves but in leading back to genesis -- i.e. DoS) or there may be other required checks to be performed when establishing a new connection with a node, vs receiving blocks from an existing connection. I would defer here: replace "it should request these immediately to sync up their chain." with "it should initiate the syncing sequence" and we'll defer details to there (as is currently done anyways)

Author: I agree with the stated problems, but am not sure how this is relevant here, but it clearly becomes cloudy with us being so specific here. I'll rephrase the last part to leave it more open to `sync` how that actually takes place. It was for sure not meant to "inject" this block into a running sync process.

### Syncing

> Whenever a node learns about a new `BlockHead` it should attempt to import the block - let that be through the `HelloMessage` or through the [block pubsub protocol](data-propagation.md#block-propagation). For that it may fetches the ancestors of that block through bitswap until it is fully caught up. Importing in this context refers to the node confirming that the block is valid, as described in [Validation](./validation.md) and storing it locally. This mode is known as "syncing". During "syncing" the same node may not mine blocks.

Reviewer: What about „caught up“ mode? Is that not a thing anymore?

Reviewer 3: typo "it may fetche**es**".

Author: @dignifiedquire I never quite understood the idea of distinct "modes", it for sure isn't how we have implemented syncing in substrate. I picture this might be a useful architecture in implementing it in certain frameworks or context, but from a spec perspective a node is a node and there aren't really distinct "modes" in which it is but just certain requirements that is has to adhere by when doing certain things. The only thing I found was the there obviously shouldn't be any "mining" while importing. But that could easily be implemented with a lock rather than understanding it as different modes. And that is much easier than thinking about all the cases of transitions between the modes that could occur IMHO.

Or is there anything in particular what I missed about these modes and that we are missing in this version?

Reviewer: Afaict there is not more to these modes, but @whyrusleeping might have some additional thoughts here, as he was the one who added them in the first place I believe.

## Establishing a network

Every node should aim to establish a set of stable connections to the network in order to stay up to date with latest changes and block propagation - aka to sync - as well as to participate in the processes of the network - e.g. mining.

In order to do that a node must be "bootstrapped" when starting up. At its simplest, bootstrapping can be thought of as secure network joining. It must enable a new node to get the latest head/chain that the majority of the network power is mining off and must be both:

- Robust — able to do so in adversarial networks (so long as honest nodes make up the majority of the network).
- Consistent — able to reliably get the canonical chain (even given poor latency or state loss in our initial peer set).

> The specific bootstrapping method and heuristic - or a mix of them - is left up to the node implementation, but we provide some example ideas of how this could be done and what aspects should be taken into consideration for that.

Reviewer 3: may be worth quickly stating the following: "should be taken into consideration for that**, such as relying on both new nodes and new connections to ensure a node is connected to both a reliable and representative slice of the network**."

Which sets up for the following sections

### Trusted bootnodes

> As with many other networks, trusted entities - the Filecoin foundation, large exchanges, good actors in the community, etc)- might publish a set of bootnodes one can connect to in order to establish network connections.

Author: replacing the ending `)` with `-`
```suggestion
As with many other networks, trusted entities - e.g. the Filecoin foundation, large exchanges, good actors in the community, etc - might publish a set of bootnodes one can connect to in order to establish network connections.
```

> Keep in mind that, because these could easily be DoS'ed, a) large set of different entities should be pre-shipped but only a few should be tried at a time b) they might drop connections rather quickly after establishing (and ban that peer for a short period of time) in order to serve more nodes. Thus you want to make sure to send a discovery request over the bootnodes to have nodes to connect to.

Reviewer 3: may be worth a reminder of what a discovery request is here

Author: In a previous discussion we agreed to not go to much into the specifics of how discovery works in this spec, as it isn't really filecoin specific nor is that a necessity for it to work. Thus the paragraph two below explains options, and this is kept vague on purpose. I added a few words to remind everyone, what a discovery is about though.

> A node should not attempt to connect to more than 10 bootnodes at a time. And should diversify its set of nodes as quickly as possible.

Reviewer 3: number seems a bit arbitrary. May change to "should not attempt to connect to more than a handful of nodes (e.g. 10) at a time: these will be used to quickly diversify a set of nodes to build a network off of"

Author: Yes, the number it sort-of arbitrary, but not newly defined. It is the combination of the [first 5 bootnodes](https://github.com/filecoin-project/specs/pull/295/files#diff-caf96f80dec81066bbd3e06dfe6b4d25L54)  and the expansion of [another 5 slots via kademlia](https://github.com/filecoin-project/specs/pull/295/files#diff-caf96f80dec81066bbd3e06dfe6b4d25L78) as previously stated.

I stated that specific number here, to allow for efficient and clearly defined spam-protection rules for anyone running the bootnodes farm: if anyone sees 10 connects over their entire cluster they are totally in their right to refuse any further connection, because that client doesn't handle according to spec. It also makes more clear the expectation to implementers that they shouldn't _and can't_ rely on just bootnodes to create their networks (we've seen that behaviour in the past, where clients wouldn't implement queries but just used bootnodes and then swamped the bootnodes and stalled the network)

Reviewer 2: +1 to picking a number for now.

A node attempts to always hold a certain amount of connections ("slots") to the network -- we recommend 25-50 on an on-the-shelf system. Whenever a new node connects, it can check their previously stored reputation or assign a default value and if that reputation is higher than the lowest currently connected nodes, may replace that connection (and thus drop the lowest quality connection) or otherwise refuse to take that connection.

In this system a node might also only record a strong decrease in reputation but not drop a connection to a peer directly even upon a strong violation, because the node may still be more useful than others. It is better to stay connected to a crappy network than to no network. However, this doesn't not free the node from still adhering to the spec itself - it should not forward said violation or its connection might still be rightously dropped. 

> We also recommend to regularly check the ranking and drop and clear up the lowest 10% of slots, leaving 5% open for incoming connections and fill up the other 5% by connecting to other nodes it can find through peer discovery.

Reviewer 3: I would pull this notion into an above section: it's a good heuristic even without the notion of "usefulness" and is worth duplicating ie, "We recommend nodes regularly diversify their peerSet by dropping a section of their peers (e.g. 10%) and making new connections."

Author: I was thinking that, too, but I am missing the background to consider it a "good heuristic". I only know this idea of dropping these connection from papers that have some form of ranking or determination of what the "lowest x%" entails. I know of "randomly drop x% of connections" only in the context of brahms, but that is a much lower number and has a much more complex structure to ensure this doesn't degrade the network over time (like certain minimal numbers of connections to always hold and a local network view to draw from).

I am not sure that generally dropping 10% of connections and _attempting_ to fill these slots with new connections (as you aren;'t guaranteed they will succeed) is not - over time - leading to you having just less and less connections until you dropped the last _good_ connection and then ... are on your own ...

> All this to creating a local view of for that node most useful connections to the network. The reputation may be stored permanently and be available between restarts - thus providing a neat bootstrap start list, too. 

Reviewer 3: typo "to **creating**"

# Filecoin Network

The filecoin network is built using libp2p building blocks, as transports and protocols, as well as some additional filecoin specific protocols as outlined in [Network Protocols](./network-protocols.md).

## Required Protocols

Every full node must support the following libp2p protocols:

 - [gossipsub](https://github.com/libp2p/specs/tree/master/pubsub/gossipsub) ([for data announcements](./data-propagation.md))
 - [bitswap] ([for data exchange](./))
 - Filecoin specific Protocols:
   - Hello Handshake
   - StorageDeal
   - BlockSync

## Transports, Streams & Encryption

> Connections between nodes are end-to-end-encrypted and -authenticated, thus every node has a `peerId` associated with it. Any node can also support unencrypted or unauthenticated connection request, but it is not required to do so.

Reviewer 2: I would leave off the bit about supporting unencrypted/unauthenticated connections. Thats a debug mode and is not to ever be used in production.

Author: > I would leave off the bit about supporting unencrypted/unauthenticated connections.

Wasn't sure if that was something we intended to support - libp2p certainly does allow that in general and I wasn't sure if that might be useful for light-client scenarios or unnecessary. I am fine with dropping it.

Reviewer 2: I think its worth taking a hardline stance on 'all connections should be encrypted and authenticated'. leaving the option open for unencrypted comms just sounds like its asking for downgrade attacks.

> Connections between nodes are end-to-end-encrypted and -authenticated, thus every node has a `peerId` associated with it. Any node can also support unencrypted or unauthenticated connection request, but it is not required to do so.

Author: ```suggestion
Connections between nodes are end-to-end-encrypted and -authenticated, thus every node has a `peerId` associated with it.
```

> Filecoin protocols are run as multiple streams over a single connection. Only one connection should exist between two nodes at any time, multiple streams within that connections must be used to exchange different protocols.

Reviewer 2: Maybe add a note at the end of this, "This is all abstracted away and handled by libp2p, Filecoin implementations with a correct libp2p library need not worry about this"

Author: I am a little hesitant to add sentences like this for two reasons: 1. with libp2p being a modular stack, it doesn't really state whether muxing is implemented, nor is it enforced by the library and 2. there is no real definition of what constitutes the "correct libp2p library"/implementation -- especially as we don't currently run cross-implementation-integration tests...

How about
```suggestion
Filecoin protocols are run as multiple streams over a single connection. Only one connection should exist between two nodes at any time, multiple streams within that connections must be used to exchange different protocols. This is usually handled from within the libp2p stack if configured properly.
```

Reviewer 2: That sounds good to me!

> This document doesn't make any assumptions on which transports and multiplexing features a particular node offers, however it must at least be able to respond and communicate through ABC AND XYZ

Author: ```suggestion
This document doesn't make any assumptions on which transports and multiplexing features a particular node offers, however it must at least be able to respond and communicate through secio using mplex channels.
```

{{% notice todo %}}

> Define minimum requirements in terms of network/e2ee-protocols and multiplexing mechanism. Should we directly jump to yamux?

Reviewer 2: We should go with secio and mplex for now, if tls 1.3 becomes mature enough all around, we could swap it for secio, but we're not there yet. Mplex also tends to be more performant, and drastically easier to implement than yamux (downside being that it doesnt have great backpressure).
